Add iterator support

JakeChampion opened the pull request:
Should solve #4

ljharb commented on implementation.js:
```diff
@@ -2,8 +2,41 @@
+		while (!done) {
+			iterableResponse = arrayLike.next();
+			if (
+				iterableResponse.hasOwnProperty('value')
```
let's use the "has" module rather than trusting `Object#hasOwnProperty`

ljharb commented on implementation.js:
```diff
@@ -2,8 +2,41 @@
+				iterableResponse.hasOwnProperty('value')
+				&& iterableResponse.hasOwnProperty('done')
+			) {
+				if (iterableResponse.done === true) {
```
Can you add comments linking to the part of the spec where `done` must be a literal true, instead of just a truthy value, and so forth?

ljharb commented on implementation.js:
```diff
@@ -2,8 +2,41 @@
+
 /*! https://mths.be/array-from v0.2.0 by @mathias */
-module.exports = function from(arrayLike) {
+module.exports = function from(arrayLike) { // eslint-disable-line complexity, max-statements
```
instead of disabling this rule, let's just bump the values

ljharb commented on implementation.js:
```diff
@@ -28,6 +61,31 @@ module.exports = function from(arrayLike) {
+	var arrayFromIterable;
+
+	// if it is an iterable treat like one
+	arrayFromIterable = parseIterable(arrayLike);
```
`var arrayFromIterable =` - no need to separate declaration from assignment

ljharb commented on implementation.js:
```diff
@@ -28,6 +61,31 @@ module.exports = function from(arrayLike) {
+		typeof arrayLike.entries === 'function'
+		&& typeof arrayLike.values === 'function'
+	) {
+		if (arrayLike.constructor.name === 'Set' && 'values' in Set.prototype) {
```
I definitely don't like the idea of using `constructor`, which is highly unreliable.

I think we need a very robust way of detecting Maps and Sets which works in every engine, at every stage of `for..of` + `Symbol.iterator` support, that users aren't capable of faking.

ljharb commented on tests/tests.js:
```diff
@@ -167,6 +188,52 @@ var runTests = function runTests(arrayFrom) {
+			}
+		}
+
+		if ('Set' in global && 'values' in Set.prototype) {
```
tests should be skipped via the `skip` option rather than if blocks - and the Map and Set tests should be different subtests of the generic "iterable" test.

ljharb commented on package.json:
```diff
@@ -1,61 +1,62 @@
-		"define-properties": "^1.1.2",
-		"es-abstract": "^1.5.0"
-	}
+  "name": "array.from",
```
please revert all the spacing changes in package.json :-/ it's always best to manually edit it instead of letting npm do it.

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,64 @@
+        return items[Symbol.iterator]();
+    } else if (isCallable(items['@@iterator'])) {
+        return items['@@iterator']();
+    } else if ('Set' in global && 'Map' in global && isNative(Set) && isNative(Map) && isCallable(items.entries) && isCallable(items.values)) {
```
why the "is native" checks? both core-js and es6-shim wrap Map and Set in most browsers.

JakeChampion replied:
I may have misunderstood your earlier comment about preventing user faking -- https://github.com/mathiasbynens/Array.from/pull/27#discussion_r68158262

ljharb replied:
Right - I'm saying that you cant use the `constructor` property, ever, for reliable anything. Not that I only want things to work with the native Map/Set.

In addition, have you tried this in browsers? There exist browsers where `Symbol.iterator` doesn't exist, and Set/Map lack `.forEach`, but `for..of` works - but `for..of` is a syntax error in older browsers.

This task is not a simple one, which is why it's still outstanding.

JakeChampion replied:
I've removed the native checks.

Not yet tested in browsers, are you willing to have browser testing to CI?

I'm not sure I understand the remarks about Set/Map lacking `forEach`, I'm not using `forEach` or `for..of`.

ljharb replied:
Sorry, I should have clarified - in these browsers, Maps and Sets lack forEach, entries, values, etc - literally the *only* way some browsers can iterate Maps and Sets is `for..of`.

Adding browser testing to CI is difficult, but sure, that'd be great. Regardless, things need to be manually tested in browsers at the least.

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,64 @@
+var usingIterator = function (items) {
+    if (hasSymbols && isCallable(items[Symbol.iterator])) {
+        return items[Symbol.iterator]();
+    } else if (isCallable(items['@@iterator'])) {
```
This fallback should only be used in browsers that ship this - namely, old versions of Firefox.

JakeChampion replied:
I've added a basic check for this

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,64 @@
+    } else if (isCallable(items['@@iterator'])) {
+        return items['@@iterator']();
+    } else if ('Set' in global && 'Map' in global && isNative(Set) && isNative(Map) && isCallable(items.entries) && isCallable(items.values)) {
+		if (items.constructor.name === 'Set') {
```
IE 9-11 don't have the "name" property on functions. At least do `items.constructor === Set`, but then of course it won't work cross-realm.

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,68 @@
+var iteratorSymbol;
+if (hasSymbols) {
+    iteratorSymbol = Symbol.iterator;
+} else if ('Set' in global && typeof new Set()['@@iterator'] === 'function') {
```
I believe it might be the case that `var nativelySupportsStringIterator = false; var s = new Set(); s.add(1); s.add(2); s['@@iterator'] = function () { nativelySupportsStringIterator = true; }); for (var x of s) { }` will be a more reliable test than just checking for the presence of a string property.

ljharb commented on package.json:
```diff
@@ -1,61 +1,64 @@
-		"define-properties": "^1.1.2",
-		"es-abstract": "^1.5.0"
-	}
+    "name": "array.from",
```
the spacing changes in this file still need to be reverted.

JakeChampion commented:
@ljharb Addressed all comments raised.

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,88 @@
 var ES = require('es-abstract/es6');
 var supportsDescriptors = require('define-properties').supportsDescriptors;
+var has = require('has');
+var global = require('global-object');
```
an alternative to this module would be https://www.npmjs.com/package/system.global which follows the official language spec proposal.

mathiasbynens replied:
+1 to using `system.global` instead.

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,88 @@
+var has = require('has');
+var global = require('global-object');
+
+var parseIterable = function (arrayLike) {
```
this argument is an iterator, not an arrayLike, no?

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,88 @@
+} else if ('Set' in global) {
+	var supportsStrIterator = (function () {
+		try {
+			return eval( // eslint-disable-line no-eval
```
at the least this should use `Function` - never `eval`.

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,88 @@
+				+ 's["@@iterator"] = function () {'
+					+ 'nativelySupportsStringIterator = true;'
+				+ '};'
+				+ 'for (var x of s) {}'
```
i believe this is the only line that requires `Function` - could we make the rest be actual code, and only enclose this line in an evalled function?

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,88 @@
+	if (iteratorSymbol && isCallable(items[iteratorSymbol])) {
+		return items[iteratorSymbol]();
+	} else if ('Set' in global && 'Map' in global && isCallable(items.entries) && isCallable(items.values)) {
+		if (items.constructor === 'Set') {
```
wouldn't the `constructor` property `=== Set`, not the *string* "Set"? (same with Map below)

ljharb commented on implementation.js:
```diff
@@ -1,9 +1,88 @@
 
 /*! https://mths.be/array-from v0.2.0 by @mathias */
-module.exports = function from(arrayLike) {
+module.exports = function from(arrayLike) { // eslint-disable-line complexity, max-statements
```
rather than disabling the linter rule, let's tweak the threshold values in `.eslintrc`.

ljharb commented on tests/tests.js:
```diff
@@ -5,6 +5,27 @@
+				};
+			} else {
+				return {
+					'done': true
```
I believe the [spec](https://tc39.github.io/ecma262/#sec-createiterresultobject) always creates a `value` property explicitly set to `undefined` - ie, `'value' in iteratorResult` must always be true.

ljharb commented on tests/tests.js:
```diff
@@ -170,6 +191,41 @@ var runTests = function runTests(arrayFrom) {
+			te.deepEqual(arrayFrom([1, 2, 3].values()), [1, 2, 3]);
+		});
+
+		test('works with Map objects', { 'skip': 'Map' in global && 'values' in Map.prototype }, function (te) {
```
this will skip the test when `Map` *is* present - i think you need to invert the condition.

ljharb commented on tests/tests.js:
```diff
@@ -170,6 +191,41 @@ var runTests = function runTests(arrayFrom) {
+			te.deepEqual(arrayFrom(map), [[1, 2], [3, 4]]);
+		});
+
+		test('works with Set objects', { 'skip': 'Set' in global && 'values' in Set.prototype }, function (te) {
```
ditto here: invert the condition.

ljharb commented on implementation.js:
```diff
@@ -1,6 +1,85 @@
 var ES = require('es-abstract/es6');
 var supportsDescriptors = require('define-properties').supportsDescriptors;
+var has = require('has');
+var global = require('system.global');
```
this actually returns a function, so you'd want to invoke it here.

mathiasbynens commented:
What’s with the failing builds?

ljharb commented:
Most of the new tests in this PR are missing explicit end or plan calls.

JakeChampion commented:
All passing now :-)

JakeChampion commented:
@ljharb This is ready for another review when you have the time. Thanks for helping with this feature by the way, I really appreciate the feedback you have given :-)

ljharb commented on implementation.js:
```diff
@@ -1,6 +1,78 @@
+	}());
+
+	if (supportsStrIterator) {
+		iteratorSymbol = '@@iterator';
```
I don't think that the above check actually indicates that it supports a string iterator - that can only be tested by setting one, and asserting that `for..of` invokes it.

ljharb commented on implementation.js:
```diff
@@ -1,6 +1,78 @@
+var isCallable = require('is-callable');
+
+var usingIterator = function (items) {
+	if (iteratorSymbol && isCallable(items[iteratorSymbol])) {
```
the spec actually says that if it's not `undefined`, it should be blindly invoked, and if that causes a TypeError, that's OK

ljharb commented on implementation.js:
```diff
@@ -1,6 +1,78 @@
+	if (iteratorSymbol && isCallable(items[iteratorSymbol])) {
+		return items[iteratorSymbol]();
+	} else if ('Set' in global && 'Map' in global && isCallable(items.entries) && isCallable(items.values)) {
+		if (items.constructor === Set) {
```
anything with `===` won't work cross-realm (like iframes, or web workers, etc) - to be robust, I think you'd need to stash a copy of the `Set.prototype.size` getter (and Map) and assert that `.call`ing on `items` did not throw.

JakeChampion replied:
I made a JSBin to test this out, it seems `Set`s and `Map`s are detectable across realms (I tested web workers) via `instanceof`. Is this what you expected?

http://jsbin.com/kiguqunoxi/edit?js,console

falsandtru replied:
Another, the following test will work on many more browsers.

```js
var iframe = document.createElement('iframe');
iframe.src = 'about:blank';
document.body.appendChild(iframe);
console.log(isMap(new iframe.contentWindow.Map()));
```

ljharb replied:
If that's the case with `instanceof`, it's a violation of the spec - the `.call` you're using is the correct way to make it work across realms.

falsandtru replied:
Yes, it is assumed to use `.call`.

ljharb commented on implementation.js:
```diff
@@ -1,6 +1,93 @@
+var iteratorSymbol;
+if (hasSymbols) {
+	iteratorSymbol = Symbol.iterator;
+} else {
```
you are still missing a third case - some versions of Safari, for instance, don't have `Symbol.iterator`, but do have Maps/Sets that are iterable only via `for..of`. In this case, we'd need to iterate over the collection via `for..of` in a dynamically created `Function`.

See https://github.com/ljharb/map-tojson/blob/master/index.js#L11-L29 for some hoops I've jumped through in the past (it might be useful to make a separate "iterate" package to encapsulate all of this logic)

JakeChampion replied:
These are the errors I've seen in iOS!

JakeChampion commented:
Addressed all comments, all passing on CI except for 3 allowed failures -- node 0.9, node 0.6, node 0.4

I'm currently running this test suite against a multitude of browsers, this usually takes a very long time. I will add a comment Wednesday morning (UK time). So far, it looks like there are errors on ios 8.2 :-/

JakeChampion commented on implementation.js:
```diff
@@ -1,6 +1,93 @@
+	if (has(items, iteratorSymbol)) {
+		return items[iteratorSymbol]();
+	} else if ('Set' in global && 'Map' in global && isCallable(items.entries) && isCallable(items.values)) {
+		if (items.constructor === Set) {
```
Won't work cross realm, need to resolve this.

JakeChampion commented on implementation.js:
```diff
@@ -1,6 +1,93 @@
+		if (items.constructor === Set) {
+			return items.values();
+		}
+		if (items.constructor === Map) {
```
Won't work cross realm, need to resolve this.

JakeChampion commented:
All browsers not finished running yet, results so far for browsers which failed:
```
- failed: <iphone 7.0 on Mac 10.9> (4 failed, 96 passed)
- failed: <iphone 7.1 on Mac 10.10> (4 failed, 96 passed)
- failed: <iphone 8.1 on Mac 10.10> (12 failed, 88 passed)
- failed: <iphone 8.2 on Mac 10.10> (12 failed, 88 passed)
- failed: <iphone 8.4 on Mac 10.10> (12 failed, 88 passed)
- failed: <iphone 9.1 on Mac 10.10> (4 failed, 96 passed)
- failed: <iphone 9.2 on Mac 10.10> (4 failed, 96 passed)
- failed: <iphone 9.0 on Mac 10.10> (4 failed, 96 passed)
- failed: <ipad 7.0 on Mac 10.9> (4 failed, 96 passed)
- failed: <ipad 7.1 on Mac 10.10> (4 failed, 96 passed)
- failed: <iphone 9.3 on Mac 10.11> (4 failed, 96 passed)
- failed: <ipad 8.2 on Mac 10.10> (12 failed, 88 passed)
- failed: <ipad 8.1 on Mac 10.10> (12 failed, 88 passed)
- failed: <ipad 8.3 on Mac 10.10> (12 failed, 88 passed)
- failed: <ipad 8.4 on Mac 10.10> (12 failed, 88 passed)
- failed: <ipad 9.0 on Mac 10.10> (4 failed, 96 passed)
- failed: <ipad 9.1 on Mac 10.10> (4 failed, 96 passed)
- failed: <ipad 9.2 on Mac 10.10> (4 failed, 96 passed)
- failed: <firefox 20 on Windows 2008> (4 failed, 96 passed)
- failed: <ipad 9.3 on Mac 10.11> (4 failed, 96 passed)
- failed: <firefox 22 on Windows 2012 R2> (4 failed, 96 passed)
- failed: <firefox 21 on Mac 10.9> (4 failed, 96 passed)
- failed: <firefox 25 on Windows 2008> (8 failed, 92 passed)
- failed: <firefox 24 on Windows 2012> (8 failed, 92 passed)
- failed: <firefox 26 on Windows 10> (8 failed, 92 passed)
- failed: <firefox 23 on Mac 10.11> (4 failed, 96 passed)
- failed: <safari 6 on Mac 10.8> (12 failed, 88 passed)
- failed: <safari 8 on Mac 10.10> (12 failed, 88 passed)
- failed: <safari 7 on Mac 10.9> (4 failed, 96 passed)
- failed: <safari 9 on Mac 10.11> (4 failed, 96 passed)
- failed: <opera 11 on Windows 2003> (4 failed, 96 passed)
- failed: <opera 12 on Windows 2003> (4 failed, 96 passed)
```

falsandtru commented on implementation.js:
```diff
@@ -1,6 +1,102 @@
+	return done ? tempArray : false;
+};
+
+var hasSymbols = typeof Symbol === 'function' && typeof Symbol.iterator === 'symbol';
```
`typeof Symbol.iterator === 'symbol'` doesn't work with Symbol polyfills.

ljharb replied:
That's because `Symbol` isn't truly polyfillable - it's a native-only feature. We should be supporting shims, but not go out of our way to support shams.

falsandtru commented on implementation.js:
```diff
@@ -1,6 +1,102 @@
+		try {
+			var supported = false;
+			var obj = { // eslint-disable-line no-unused-vars
+				'@@iterator': function () {
```
should be `[Symbol.iterator]: function () {`

ljharb replied:
This is specifically handling the old Firefox edge case where they supported the string "@@iterator".

falsandtru replied:
Thanks, I didn't know that.

falsandtru commented on implementation.js:
```diff
@@ -1,6 +1,102 @@
+var usingIterator = function (items) {
+	if (has(items, iteratorSymbol)) {
+		return items[iteratorSymbol]();
+	} else if ('Set' in global && 'Map' in global && isCallable(items.entries) && isCallable(items.values)) {
```
probably special cases are not needed

ljharb replied:
Indeed they are needed, because in some engines, there are no Symbols, no methods that produce iterators, but there is Set/Map, and they're iterable with `for..of`.

falsandtru replied:
I see, thanks.

falsandtru commented:
I refined this support in #30. It would work better. Could you review it?

JakeChampion commented:
I ran the tests against the master using my list of browsers, it seems to have the same 4 failing tests.

JakeChampion commented:
```
   - failed: <iphone 7.0 on Mac 10.9> (4 failed, 80 passed)
    1 browser(s) failed
  Failed Tests: There were 4 failures
    no setters are called for indexes
      ✖ should throw
    no setters are called for indexes
      ✖ should throw
    no setters are called for indexes
      ✖ should throw
    no setters are called for indexes
      ✖ should throw
  total:     261
  passing:   257
  failing:   4
  duration:  1m 15.1s
~/Code/oss/Array.from master 1m 15s
```

JakeChampion commented:
I've commented out the test which is failing on master and am now rerunning against all the browsers on SauceLabs. I have high hopes it will work :-)

ljharb commented:
Hmm - I'd like to fix those tests on master before merging this.

JakeChampion commented:
🎉 🎉 🎉 🎉 🎉 🎉 🎉 
```
- passed: <iphone 7.0 on Mac 10.9>
- passed: <iphone 7.1 on Mac 10.10>
- passed: <iphone 9.0 on Mac 10.10>
- passed: <iphone 9.1 on Mac 10.10>
- passed: <iphone 9.2 on Mac 10.10>
- passed: <iphone 9.3 on Mac 10.11>
- passed: <ipad 7.0 on Mac 10.9>
- passed: <ipad 7.1 on Mac 10.10>
- passed: <ipad 9.0 on Mac 10.10>
- passed: <ipad 9.1 on Mac 10.10>
- passed: <ipad 9.2 on Mac 10.10>
- passed: <ipad 9.3 on Mac 10.11>
- passed: <firefox 4 on Windows 10>
- passed: <firefox 5 on Windows 2008>
- passed: <firefox 6 on Mac 10.8>
- passed: <firefox 7 on Windows 10>
- passed: <firefox 8 on Mac 10.11>
- passed: <firefox 9 on Windows 2012 R2>
- passed: <firefox 10 on Mac 10.8>
- passed: <firefox 11 on Windows 2008>
- passed: <firefox 12 on Windows 2012>
- passed: <firefox 45 on Windows 10>
- passed: <firefox 46 on Mac 10.11>
- passed: <firefox 47 on Mac 10.11>
- passed: <firefox 48 on Windows 2012>
- passed: <safari 5 on Windows 2008>
- passed: <safari 7 on Mac 10.9>
- passed: <safari 9 on Mac 10.11>
- passed: <chrome 26 on Windows 10>
- passed: <chrome 27 on Linux>
- passed: <chrome 28 on Windows 2003>
- passed: <chrome 29 on Windows 10>
- passed: <chrome 50 on Mac 10.9>
- passed: <chrome 51 on Mac 10.10>
- passed: <chrome 52 on Mac 10.10>
- passed: <internet explorer 9 on Windows 2008>
- passed: <internet explorer 10 on Windows 2012>
- passed: <microsoftedge 13 on Windows 10>
- passed: <opera 11 on Windows 2003>
- passed: <opera 12 on Windows 2003>
- passed: <android 4.0 on Linux>
- passed: <android 4.1 on Linux>
- passed: <android 4.3 on Linux>
- passed: <android 4.2 on Linux>
- passed: <android 4.4 on Linux>
- passed: <android 5.0 on Linux>
- passed: <android 5.1 on Linux>
```

JakeChampion commented:
@mathiasbynens @ljharb I believe this is ready for it's final review :-)

ljharb commented:
@mathiasbynens btw would you mind enabling required PR approvals on this repo, in settings? :-)

@JakeChampion thanks! I'll take a final look and merge tonight.

ljharb commented:
Thanks! I'm going to give this a couple days to bake, and for me to test and play around, before releasing it.

le0nik commented:
Iterators still don't work in `latest` as it is `1.0.3`. Should users just install master until this is released?

mroderick commented:
Or use https://github.com/studio-b12/array-from until a new version is published to NPM
